src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala

```diff
@@ -183,6 +186,23 @@ class BlockFetcher(
       fetchBlocks(newState)
   }
 
+  private def handlePossibleTopUpdate(state: BlockFetcherState): Receive = {
+    //by handling these type of messages, fetcher can received from network, fresh info about blocks on top
+    //ex. After a successful handshake, fetcher will receive the info about the header of the peer best block
+    case MessageFromPeer(BlockHeaders(headers), _) =>
+      headers.lastOption.map { bh =>
+        log.debug(s"Candidate for new top at block ${bh.number}, current know top ${state.knownTop}")
+        val newState = state.withPossibleNewTopAt(bh.number)
+        fetchBlocks(newState)
+      }
+    //keep fetcher state updated in case new checkpoint block or mined block was imported
+    case LastBlockChanged(blockNr) => {
+      log.debug(s"New last block $blockNr imported from the inside")
+      val newState = state.withLastBlock(blockNr).withPossibleNewTopAt(blockNr)
```

Comment:
Last block seems to also involve the [last headers fetched](https://github.com/input-output-hk/mantis/blob/70c8dc5068066e9a04cbea82d77e505ccb91fb1b/src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcherState.scala#L60), what if we are already in progress of fetching blocks from the network with block number higher than the one imported (as it's likely the case for checkpoint blocks)? Doesn't this break the consistency of the fetcher?

Maybe the new last block should be `blockNr.max(state.lastBlock)` or sth like that

Reply:
good catch